frontend: refactor password.jsx into typescript

NicolaLS opened the pull request:
Refactor the `password.jsx` file into typescript.

I had to cast to the concrete type a lot of times, I'm not sure if that makes sense since it can throw an error, but how _should_ an error be handled? e.g. `this.setState({ [stateKey]: value } as Pick<TState, keyof TState>, this.validate);`

thisconnect commented on frontends/web/src/routes/device/bitbox01/unlock.jsx:
```diff
@@ -135,12 +135,10 @@ class Unlock extends Component {
                         <PasswordSingleInput
                           autoFocus
                           id="password"
-                          type="password"
```
👍

thisconnect commented on frontends/web/src/routes/device/components/skipfortesting.tsx:
```diff
@@ -58,11 +58,9 @@ export const SkipForTesting = ({
             label="Test Password"
-            onValidPassword={setTestPIN}
-            value={testPIN} />
+            onValidPassword={(pw) => pw ? setTestPIN(pw) : setTestPIN('')}/>
```
with typing this you found a small bug here. 👍

thisconnect commented on frontends/web/src/components/password.tsx:
```diff
@@ -302,9 +340,10 @@ class PasswordRepeatInputClass extends Component {
+export { HOCRepeat as PasswordRepeatInput };
 
-const MatchesPattern = ({ regex, value = '', text }) => {
+const MatchesPattern = ({ regex, value = '', text }: { regex: RegExp | undefined, value: string, text: string | undefined }) => {
```
please change to `type MatchesPatternProps = {`

and use MatchesPatternProps here.

thisconnect commented on frontends/web/src/components/password.tsx:
```diff
@@ -316,7 +355,7 @@ const MatchesPattern = ({ regex, value = '', text }) => {
 const excludeKeys = /^(Shift|Alt|Backspace|CapsLock|Tab)$/i;
 
-const hasCaps = ({ key }) => {
+const hasCaps = ({ key }: { key: string }) => {
```
not sure if better to use `{ key: string }` or just `KeyboardEvent`

thisconnect replied:
actually there is another bug with `event.shiftKey` at the end of this function, please remove both `@ts-ignore` and `eslint-disable-next-line` and check for a solution 🙏

NicolaLS replied:
ah I think the whole function can just return `event.getModifierState('CapsLock')` then, the bug with that always returning false on qt was probably because `event` was not globally available and it was also not passed to the function.

Actually we can remove the function completely if `event.getModifierState` works, will test on qt now.

NicolaLS replied:
Looks like adding the event listener does not even work when it is done in `UNSAFE_componentWilMount`! I'll remove the function and add it to `componentDidMount`

I noticed I can't really test the input dialog on qt, but I'll put the same logic in some other component just to test.

NicolaLS replied:
Ok so `event.getModifierState('CapsLock')` still does not work on qt linux.

thisconnect replied:
yeah I believe @shonsirsha mentioned something like that before..

thisconnect commented on frontends/web/src/components/password.tsx:
```diff
@@ -211,18 +249,18 @@ class PasswordRepeatInputClass extends Component {
+  handleCheckCaps = (event: KeyboardEvent) => {
     const capsLock = hasCaps(event);
-        if (capsLock != null) { // eslint-disable-line
+    if (capsLock != null) { // eslint-disable-line
```
since you are here and this is now typed, it looks like ` // eslint-disable-line` isnt needed anymore
```suggestion
    if (capsLock !== null) {
```

NicolaLS commented on frontends/web/src/components/password.tsx:
```diff
@@ -88,16 +120,16 @@ class PasswordSingleInputClass extends Component {
-    this.setState({ [stateKey]: value }, this.validate);
+    const stateKey = event.target.id.slice(this.idPrefix().length) as keyof TState;
+    this.setState({ [stateKey]: value } as Pick<TState, keyof TState>, this.validate);
   };
```
I'll take a note and either open a PR or issue.

NicolaLS commented:
@thisconnect thanks, PTAL.

note: I removed the `UNSAFE_componentWillMount` and added the `keydown` event listener in the `componentDidMount` now because (at least on `webdev` firefox) the listener did not work (the method ran, but the listener was either not added correctly or broken).

thisconnect commented on frontends/web/src/components/password.tsx:
```diff
@@ -302,9 +349,15 @@ class PasswordRepeatInputClass extends Component {
+type MatchesPatternProps = {
+  regex: RegExp | undefined,
+  value: string,
+  text: string | undefined,
```
here and in other type in this file `,` -> `;`

thisconnect replied:
ptal

thisconnect replied:
PTAL 🙏

thisconnect replied:
self-review the PR after you push then you see :)

thisconnect commented on frontends/web/src/components/password.tsx:
```diff
@@ -153,29 +195,40 @@ class PasswordSingleInputClass extends Component {
+export { HOC as PasswordSingleInput };
+
 
+type TPasswordRepeatProps = TPasswordSingleInputProps & { repeatLabel?: string; repeatPlaceholder: string; };
```
please use newlines for readabilty

thisconnect reviewed (changes requested):
sorry for the slow review. looks all good with a typescript nits in 2 places.

please update then I do a final test with BB01

NicolaLS commented:
@thisconnect thanks, first push was the actual change, last push is a rebase on master.

NicolaLS commented on frontends/web/src/components/password.tsx:
```diff
@@ -60,8 +108,9 @@ class PasswordSingleInputClass extends Component {
-  tryPaste = event => {
-    if (event.target.type === 'password') {
+  tryPaste = (event: React.ClipboardEvent<HTMLInputElement>) => {
+    const target = event.target as HTMLInputElement;
```
oh makes sense, I did not understand why _typescript was so stupid_ but I used the wrong property..

thisconnect reviewed (approved):
tested LGTM
